# Exercise 4.8

## The Script contains the following points:

    1. Importing Libraries
    2. Importing Dataframes
    3. Finding the aggregated mean of the 'order_number' column grouped by 'department_id'
    4. Analysis and how the results differ
    5. Creating the loyalty flag for existing customers
    6. Determine whether the prices of products purchased by local custers differ from those purchased by regular
       or new customers
    7. Creating an order frequency flag 
    8. Exporting Data

## Importing Libraries 

In [1]:
# Important Libraries 
import pandas as pd
import numpy as np
import os

## Importing Dataframes

In [2]:
path = r'/Users/datcaodo/Documents/22-03-2024 Instacart Basket Analysis'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

## Question 2 - Finding the aggregated mean of the 'order_number' column grouped by 'department_id'

In [4]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Question 3 - Analysis and how the results differ from the subset

The results differ because in the subset we only use the first 1000000 rows whereas in this analysis it is the whole dataframe which will come out with a different mean. Therefore the mean of the order_number of the mean differ.

## Question 4 - Creating the loyalty flag for existing customers

In [5]:
#Splitting the data in 'user_id'
ords_prods_merge.groupby('user_id')

In [6]:
#Creating a new column, 'max_order'
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/kq/b6pkx345701cskyhyz_88fhm0000gp/T/ipykernel_4066/2571959081.py:2: FutureWarning: The provided callable <function max at 0x1114919e0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [7]:
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  orders_day_of_week  \
0    3139998      138    prior            28                   6   
1    1977647      138    prior            30                   6   
2     389851      709    prior             2                   0   
3     652770      764    prior             1                   3   
4    1813452      764    prior             3                   4   
5    1701441      777    prior            16                   1   
6    1871483      825    prior             3                   2   
7    1290456      910    prior            12                   3   
8     369558     1052    prior            10                   1   
9     589712     1052    prior            15                   1   
10   1798115     1379    prior             1                   5   
11    497193     1480    prior            36                   3   
12   1727826     1480    prior            38                   4   
13   2694698     1480    prior            39                   2   
14   3171753     1494    prior             6                   3   

    order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                  11                     3.0                  5          0   
1                  17                    20.0                  1          1   
2                  21                     6.0                 20          0   
3                  13                     NaN                 10          0   
4                  17                     9.0                 11          1   
5                   7                    26.0                  7          0   
6                  14                    30.0                  2          0   
7                  10                    30.0                  1          0   
8                  20                    19.0                  1          0   
9                  12                    15.0                  2          1   
10                 11                     NaN                  1          0   
11                 13                     7.0                  1          0   
12                 13                     7.0                  3          1   
13                 12                    12.0                  4          1   
14                 10                     2.0                  8          0   

   _merge    price_range_loc     busiest_day    busiest_days  \
0    both  Mid-range product  Regularly busy  Regularly busy   
1    both  Mid-range product  Regularly busy  Regularly busy   
2    both  Mid-range product     Busiest day    Busiest days   
3    both  Mid-range product  Regularly busy    Slowest busy   
4    both  Mid-range product      Least busy    Slowest busy   
5

In [8]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [9]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [10]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

We do not have almost 16 million (15876776) Regular customers, almost 10 million (10284093) Loyal customers, and almost 6 million (6243990) New customers.

In [27]:
# That's why we should count unique (Distinct) values of user_id in each group: 

ords_prods_merge.groupby(['loyalty_flag']).user_id.nunique() 

# here each user is counted only once

loyalty_flag
Loyal customer       17017
New customer        112328
Regular customer     76864
Name: user_id, dtype: int64

In [12]:
# View updated dataframe
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998      138    prior            28                   6  ...   
1   1977647      138    prior            30                   6  ...   
2    389851      709    prior             2                   0  ...   
3    652770      764    prior             1                   3  ...   
4   1813452      764    prior             3                   4  ...   

   days_since_prior_order  add_to_cart_order  reordered  _merge  \
0                     3.0                  5          0    both   
1                    20.0                  1          1    both   
2                     6.0                 20          0    both   
3                     NaN                 10          0    both   
4                     9.0                 11          1    both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most orders   
1  Mid-range product  Regularly busy  Regularly busy        Average orders   
2  Mid-range product     Busiest day    Busiest days        Average orders   
3  Mid-range product  Regularly busy    Slowest busy           Most orders   
4  Mid-range product      Least busy    Slowest busy        Average orders   

  max_order      loyalty_flag  
0        32  Regular customer  
1        32  Regular customer  
2         5      New customer  
3         3      New customer  
4         3      New customer  

[5 rows x 21 columns]

## Question 5 - Determine whether the prices of products purchased by local custers differ from those purchased by regular or new customers

In [13]:
# Use groupby and agg() to determine basic statisitcs of spending habits by loyalty group
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

The average price of a new customer is about 3 dollars more than of a loyal customer. 

## Question 6 - Creating a spending flag 

In [14]:
#Use transform() to determine the mean price of products purchased by each user
ords_prods_merge['mean_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/kq/b6pkx345701cskyhyz_88fhm0000gp/T/ipykernel_4066/1592812295.py:2: FutureWarning: The provided callable <function mean at 0x1114923e0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['mean_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [15]:
# View relevant columns 
ords_prods_merge[['user_id', 'mean_product_price']].head(20)

user_id  mean_product_price
0       138            6.935811
1       138            6.935811
2       709            7.930208
3       764            4.972414
4       764            4.972414
5       777            6.935398
6       825            5.957576
7       910            6.680000
8      1052            7.162500
9      1052            7.162500
10     1379            5.813636
11     1480            5.081633
12     1480            5.081633
13     1480            5.081633
14     1494            6.803797
15     1494            6.803797
16     1494            6.803797
17     1540            6.607122
18     1540            6.607122
19     1540            6.607122

In [16]:
#Create a new spending_flag column
ords_prods_merge.loc[ords_prods_merge['mean_product_price'] < 10, 'spending_flag'] = 'Low Spender'

In [17]:
ords_prods_merge.loc[ords_prods_merge['mean_product_price'] >= 10, 'spending_flag'] = 'High Spender'

In [18]:
#Check frequency of new spending_flag column
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low Spender     31770614
High Spender      634245
Name: count, dtype: int64

In [19]:
# View relevant columns of updated data frame
ords_prods_merge[['user_id', 'mean_product_price', 'spending_flag']].head(20)

user_id  mean_product_price spending_flag
0       138            6.935811   Low Spender
1       138            6.935811   Low Spender
2       709            7.930208   Low Spender
3       764            4.972414   Low Spender
4       764            4.972414   Low Spender
5       777            6.935398   Low Spender
6       825            5.957576   Low Spender
7       910            6.680000   Low Spender
8      1052            7.162500   Low Spender
9      1052            7.162500   Low Spender
10     1379            5.813636   Low Spender
11     1480            5.081633   Low Spender
12     1480            5.081633   Low Spender
13     1480            5.081633   Low Spender
14     1494            6.803797   Low Spender
15     1494            6.803797   Low Spender
16     1494            6.803797   Low Spender
17     1540            6.607122   Low Spender
18     1540            6.607122   Low Spender
19     1540            6.607122   Low Spender

## Question 7 - Creating an order frequency flag 

In [20]:
# Use groupby and transform to determine median number of days between orders for each user
ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/kq/b6pkx345701cskyhyz_88fhm0000gp/T/ipykernel_4066/2923105074.py:2: FutureWarning: The provided callable <function median at 0x1115dae80> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


## Create a new frequency_flag column

In [21]:
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'frequency_flag'] = 'Non-frequent Customer'

In [22]:
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] > 10) & (ords_prods_merge['median_days_between_orders']<= 20) , 'frequency_flag'] = 'Regular Customer'

In [23]:
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [24]:
# Check value counts of new order_frequency column
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        21559853
Regular Customer          7208564
Non-frequent Customer     3636437
nan                             5
Name: count, dtype: int64

In [25]:
# View relevant columns of updated data frame
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'frequency_flag']].head(25)

user_id  days_since_prior_order  median_days_between_orders  \
0       138                     3.0                         8.0   
1       138                    20.0                         8.0   
2       709                     6.0                         8.0   
3       764                     NaN                         9.0   
4       764                     9.0                         9.0   
5       777                    26.0                        11.0   
6       825                    30.0                        20.0   
7       910                    30.0                         6.0   
8      1052                    19.0                        10.0   
9      1052                    15.0                        10.0   
10     1379                     NaN                        20.0   
11     1480                     7.0                         7.0   
12     1480                     7.0                         7.0   
13     1480                    12.0                         7.0   
14     1494                     2.0                         7.0   
15     1494                     6.0                         7.0   
16     1494                    10.0                         7.0   
17     1540                     7.0                         7.0   
18     1540                     6.0                         7.0   
19     1540                    14.0                         7.0   
20     1540                     8.0                         7.0   
21     1540                     6.0                         7.0   
22     1540                    28.0                         7.0   
23     1540                     7.0                         7.0   
24     1540                     7.0                         7.0   

       frequency_flag  
0   Frequent Customer  
1   Frequent Customer  
2   Frequent Customer  
3   Frequent Customer  
4   Frequent Customer  
5    Regular Customer  
6    Regular Customer  
7   Frequent Customer  
8   Frequent Customer  
9   Frequent Customer  
10   Regular Customer  
11  Frequent Customer  
12  Frequent Customer  
13  Frequent Customer  
14  Frequent Customer  
15  Frequent Customer  
16  Frequent Customer  
17  Frequent Customer  
18  Frequent Customer  
19  Frequent Customer  
20  Frequent Customer  
21  Frequent Customer  
22  Frequent Customer  
23  Frequent Customer  
24  Frequent Customer

## Exporting Data

In [28]:
# Export as pickle file 
# Export derived data frame as pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_grouped.pkl'))